GSoC/2016/JuliaQuantum-:-Framework-for-QuComput.md

```diff
@@ -0,0 +1,420 @@
+# let's add a Hadamard gate to the first qubit in this three-qubit circuit
+
+# first create your circuit
+c = Circuit()
+addgate!(c,Hadamard,1)
+
+# for controlled gates
+# |a> -----x--------
+#          |
+# |b> ----[H]-------
+# |c> --------------
+
+c_H = Circuit()
+addgate!(c_H, 1, Hadamard, 2)
+# or for multiple input gates(will be useful for other gate-like modules)
+
+# |a> --------x------
+#             |
+# |b> ------|***|----
+#           | M |
+# |c> ------|***|----
+c_M = Circuit()
+addgate!(c_M, 1, M, 2, 3)
+
+# or for multiple controlled gates
+
+# |a> -------x--------
+#            |
+# |b> -------x--------
+#            |
+# |c> ------[H]-------
+cc_H = Circuit()
+addgate!(cc_H, [1,2], Hadamard, 3)
+
+```
+
+to measure a circuit, simply use the `measure` method on the circuit you want to measure with the the circuit's input state
+
+```julia
+# measure the circuit:
+# |a> ----[H]----
+# |b> -----------
+# |c> -----------
+# with input state:
+# 1/2*sqrt(2)(|000>+|001>+|010>+...+|111>)
+
+input_state = InitState(3)
+circuit = Circuit()
+addgate!(circuit, Hadamard, 1)
+res_state = measure(circuit,input_state)
+```
+
+Remove a gate from circuit
+```julia
+# given a circuit
+# |a> ----[H]-------
+# |b> ---------[H]--
+# |c> --------------
+
+removegate!(circuit,1)
+
+# remove the first gate
+# |a> --------------
+# |b> --------[H]---
+# |c> --------------
+```
+
+Prototype 2: quantum  Fourier Transformation
+![QFT circuit pic](imag/fourier.svg)
+
+```julia
+# Quantum Fourier Transformation
+init_state = InitStates(4)# inputs the number of bits
+
+QFT_curcuit = Circuit()
+
+for i=1:4
+    addgate!(QFT_circuit,Hadamard,i)
+    for j=1:i-1
+        addgate!(QFT_circuit,bitID-i,R_k(i),1)
+    end
+end
+
+res_state = measure(QFT_circuit,init_state)
+# suppose this to be the command line(the result is made up)
+---
+measurement of QFT_circuit:
+states| prob
+|0000>: 0.0001
+|0001>: 0.2000
+...
+|1111>: 0.2000
+---
+
+figure(1)
+# For visualization, a circuit should be plotted by
+plot(QFT_circuit)
+show()
+```
+
+Prototype 3:Daemon-like cooling
+
+![cooling circuit](imag/circuit.png)
+
+*This picture comes from [doi:10.103](http://www.nature.com/nphoton/journal/v8/n2/full/nphoton.2013.354.html)*
+
+```julia
+# Daemon-like Cooling Circuit
+# This prototype aims to show how other computation should be inserted in this framework
+
+init_state = PrepareBellStates(4)
+cooling = Circuit()
+
+addgate(cooling,1,"Hadamard")
+#if the gate needs parameters,will be passed after its name
+addgate(cooling,1,"R",gamma) 
+addgate(cooling,1,[2,3,4],"TimeEvo",Hs,t)
+addgate(cooling,1,"Hadamard")
+
+res_state = measure(cooling)
+
+figure(1)
+#output figure will looks like the circuit figure in the picture above
+plot(cooling)
+show()
+```
+
+Prototype 4: one-way model
+
+```julia
+adj = [1 0;
+       0 1]
+
+state = InitState(adj) #will optimize the structure of graph to minimum resource cost structure automatically
+```
+
+# Proposal Timeline
+
```

Comment:
I guess you could split the timeline into 3 parts : 
1. (Before) Community Bonding Period, 
2. Coding period (split into 2 parts : 
    i. Part - 1 : From the start to Mid-term evaluations 
   ii. Part - 2 : From Mid-term evaluations to End Term)
3. Post GSoC 

For the coding part split the major target into minor targets which could be achieved like in fewer weeks. So for example, the timeline would like : 

Week 1 - Week 3 : 
A list of TODO tasks
Targets to be achieved : Summary of the smaller targets achieved  

Similarly for other weeks !
Splitting the timeline would give everyone much better clarity on the targets of the project. Also it would give you more clarity where you might require help and assistance !